File: backend/app/repositories/reservations.py
"""Equivalente a repositories/reservations.ts.

Toda leitura de reserva já traz nome/preço do evento e o número do assento (o que a API sempre
precisa), evitando N+1 — mesmo espírito do JOIN feito no original. `ReservationView` é o
equivalente ao tipo `Reservation` "achatado" de types.ts.
"""
from dataclasses import dataclass
from datetime import datetime

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Event, Reservation, Ticket


@dataclass
class ReservationView:
    id: int
    reservation_token: str
    event_id: int
    ticket_id: int
    user_id: str | None
    user_email: str | None
    expires_at: datetime
    status: str
    created_at: datetime
    updated_at: datetime
    event_name: str
    event_price: str
    seat_number: str


def _select_joined():
    return select(
        Reservation.id,
        Reservation.reservation_token,
        Reservation.event_id,
        Reservation.ticket_id,
        Reservation.user_id,
        Reservation.user_email,
        Reservation.expires_at,
        Reservation.status,
        Reservation.created_at,
        Reservation.updated_at,
        Event.name.label("event_name"),
        Event.price.label("event_price"),
        Ticket.seat_number,
    ).join(Event, Event.id == Reservation.event_id).join(Ticket, Ticket.id == Reservation.ticket_id)


def _to_view(row) -> ReservationView:
    return ReservationView(
        id=row.id,
        reservation_token=row.reservation_token,
        event_id=row.event_id,
        ticket_id=row.ticket_id,
        user_id=row.user_id,
        user_email=row.user_email,
        expires_at=row.expires_at,
        status=row.status,
        created_at=row.created_at,
        updated_at=row.updated_at,
        event_name=row.event_name,
        event_price=str(row.event_price),
        seat_number=row.seat_number,
    )


async def find_by_token(session: AsyncSession, token: str) -> ReservationView | None:
    result = await session.execute(_select_joined().where(Reservation.reservation_token == token))
    row = result.first()
    return _to_view(row) if row else None


async def find_by_id_for_update(session: AsyncSession, reservation_id: int) -> ReservationView | None:
    # FOR UPDATE OF r: trava só a linha da reserva (evento e assento são travados à parte, em ordem fixa).
    result = await session.execute(
        _select_joined().where(Reservation.id == reservation_id).with_for_update(of=Reservation)
    )
    row = result.first()
    return _to_view(row) if row else None


async def exists_active_for_ticket(session: AsyncSession, ticket_id: int) -> bool:
    result = await session.execute(
        select(Reservation.id).where(Reservation.ticket_id == ticket_id, Reservation.status == "ACTIVE").limit(1)
    )
    return result.first() is not None


async def insert(
    session: AsyncSession,
    *,
    reservation_token: str,
    event_id: int,
    ticket_id: int,
    user_id: str | None,
    user_email: str | None,
    expires_at: datetime,
) -> ReservationView:
    reservation = Reservation(
        reservation_token=reservation_token,
        event_id=event_id,
        ticket_id=ticket_id,
        user_id=user_id,
        user_email=user_email,
        expires_at=expires_at,
        status="ACTIVE",
    )
    session.add(reservation)
    await session.flush()
    view = await find_by_id_for_update(session, reservation.id)
    if view is None:
        raise RuntimeError("INSERT de reserva não retornou a linha criada")
    return view


async def update_status(session: AsyncSession, reservation_id: int, status: str) -> None:
    reservation = await session.get(Reservation, reservation_id)
    if reservation is not None:
        reservation.status = status


async def update_expires_at(session: AsyncSession, reservation_id: int, expires_at: datetime) -> bool:
    """Só atualiza se a reserva ainda estiver ACTIVE; devolve se atualizou."""
    reservation = await session.get(Reservation, reservation_id)
    if reservation is None or reservation.status != "ACTIVE":
        return False
    reservation.expires_at = expires_at
    return True


async def find_expired_active(session: AsyncSession, now: datetime) -> list[ReservationView]:
    result = await session.execute(
        _select_joined().where(Reservation.status == "ACTIVE", Reservation.expires_at < now).order_by(Reservation.expires_at)
    )
    return [_to_view(row) for row in result.all()]

File: src/lib/repositories/reservations.ts
import type { Reservation, ReservationStatus } from '../types';
import type { ReservationRepository } from './interfaces';

type ReservationRow = {
  id: number;
  reservation_token: string;
  event_id: number;
  ticket_id: number;
  user_id: string | null;
  user_email: string | null;
  expires_at: Date;
  status: ReservationStatus;
  created_at: Date;
  updated_at: Date;
  event_name: string;
  event_price: string;
  seat_number: string;
};

// Toda leitura de reserva já traz nome/preço do evento e o número do assento (o que a API sempre
// precisa), evitando o carregamento lazy do JPA. O alias `r` é a tabela ou o CTE de INSERT.
const JOINED_COLUMNS = `
  r.id, r.reservation_token, r.event_id, r.ticket_id, r.user_id, r.user_email, r.expires_at, r.status,
  r.created_at, r.updated_at, e.name AS event_name, e.price AS event_price, t.seat_number`;
const JOINS = 'JOIN events e ON e.id = r.event_id JOIN tickets t ON t.id = r.ticket_id';

// Datas vão como ISO-8601 em UTC: colunas TIMESTAMP ignoram o sufixo "Z", e assim o valor gravado
// não depende do fuso horário do processo Node.
const iso = (date: Date): string => date.toISOString();

function toReservation(row: ReservationRow): Reservation {
  return {
    id: row.id,
    reservationToken: row.reservation_token,
    eventId: row.event_id,
    ticketId: row.ticket_id,
    userId: row.user_id,
    userEmail: row.user_email,
    expiresAt: row.expires_at,
    status: row.status,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
    eventName: row.event_name,
    eventPrice: row.event_price,
    seatNumber: row.seat_number,
  };
}

export const reservationRepository: ReservationRepository = {
  async findByToken(q, token) {
    const res = await q.query<ReservationRow>(
      `SELECT ${JOINED_COLUMNS} FROM reservations r ${JOINS} WHERE r.reservation_token = $1`,
      [token],
    );
    const row = res.rows[0];
    return row ? toReservation(row) : null;
  },

  async findByIdForUpdate(q, id) {
    // FOR UPDATE OF r: trava só a linha da reserva (evento e assento são travados à parte, em ordem fixa).
    const res = await q.query<ReservationRow>(
      `SELECT ${JOINED_COLUMNS} FROM reservations r ${JOINS} WHERE r.id = $1 FOR UPDATE OF r`,
      [id],
    );
    const row = res.rows[0];
    return row ? toReservation(row) : null;
  },

  async existsActiveForTicket(q, ticketId) {
    const res = await q.query<{ exists: boolean }>(
      `SELECT EXISTS (SELECT 1 FROM reservations WHERE ticket_id = $1 AND status = 'ACTIVE') AS "exists"`,
      [ticketId],
    );
    return res.rows[0]?.exists === true;
  },

  async insert(q, reservation) {
    const res = await q.query<ReservationRow>(
      `WITH r AS (
         INSERT INTO reservations (reservation_token, event_id, ticket_id, user_id, user_email, expires_at, status)
         VALUES ($1, $2, $3, $4, $5, $6, 'ACTIVE')
         RETURNING *
       )
       SELECT ${JOINED_COLUMNS} FROM r ${JOINS}`,
      [
        reservation.reservationToken,
        reservation.eventId,
        reservation.ticketId,
        reservation.userId,
        reservation.userEmail,
        iso(reservation.expiresAt),
      ],
    );
    const row = res.rows[0];
    if (!row) throw new Error('INSERT de reserva não retornou a linha criada');
    return toReservation(row);
  },

  async updateStatus(q, id, status) {
    await q.query('UPDATE reservations SET status = $2 WHERE id = $1', [id, status]);
  },

  async updateExpiresAt(q, id, expiresAt) {
    const res = await q.query(`UPDATE reservations SET expires_at = $2 WHERE id = $1 AND status = 'ACTIVE'`, [id, iso(expiresAt)]);
    return res.rowCount === 1;
  },

  async findExpiredActive(q, now) {
    const res = await q.query<ReservationRow>(
      `SELECT ${JOINED_COLUMNS} FROM reservations r ${JOINS}
       WHERE r.status = 'ACTIVE' AND r.expires_at < $1
       ORDER BY r.expires_at`,
      [iso(now)],
    );
    return res.rows.map(toReservation);
  },
};
